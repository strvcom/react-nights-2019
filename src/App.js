import React, { Component } from 'react'
import { Switch, Route } from 'react-router-dom'

import GlobalStyles from './globalStyles'

import { ProductList } from './pages/ProductList'
import { ProductDetail } from './pages/ProductDetail'

class App extends Component {
  render() {
    return (
      <React.Fragment>
        <GlobalStyles />
        <Switch>
          <Route path="/" exact component={ProductList} />
          <Route path="/:productId" component={ProductDetail} />
        </Switch>
      </React.Fragment>
    )
  }
}

export default App
