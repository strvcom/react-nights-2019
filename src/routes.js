import { generatePath } from 'react-router'

export const HOMEPAGE = '/'
export const PRODUCT_LIST = '/products'
export const PRODUCT_DETAIL = '/products/:productId'
export const getProductDetailRoute = productId =>
  generatePath(PRODUCT_DETAIL, { productId })
export const CART = '/cart'
export const SIGN_UP = '/signup'
export const LOGIN = '/login'
export const ACCOUNT = '/account'
export const LOGOUT = '/logout'
