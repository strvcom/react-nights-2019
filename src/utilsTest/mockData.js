const FIRST_PRODUCT = {
  id: 1,
  attributes: {
    name: 'First Test Product',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam pellentesque in neque vitae tincidunt. In gravida eu ipsum non condimentum. Curabitur libero leo, gravida a dictum vestibulum, sollicitudin vel quam.',
    image_url: 'https://img.commercelayer.io/skus/1.png',
  },
  relationships: {
    prices: {
      data: [{ id: 1 }],
    },
  },
}

const SECOND_PRODUCT = {
  id: 2,
  attributes: {
    name: 'Second Test Product',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam pellentesque in neque vitae tincidunt. In gravida eu ipsum non condimentum. Curabitur libero leo, gravida a dictum vestibulum, sollicitudin vel quam.',
    image_url: 'https://img.commercelayer.io/skus/1.png',
  },
  relationships: {
    prices: {
      data: [{ id: 2 }],
    },
  },
}

const INCLUDES = [
  {
    id: 1,
    attributes: {
      amount_cents: 1000,
      amount_float: 10,
      formatted_amount: '$10',
    },
  },
  {
    id: 2,
    attributes: {
      amount_cents: 2000,
      amount_float: 20,
      formatted_amount: '$20',
    },
  },
]

export const PRODUCTS = {
  data: [FIRST_PRODUCT, SECOND_PRODUCT],
  included: INCLUDES,
  meta: {
    page_count: 1,
    record_count: 2,
  },
}

export const PRODUCT = {
  data: FIRST_PRODUCT,
  included: INCLUDES,
}

export const USER = {
  id: 1,
  email: 'michal@test.com',
  firstName: 'Michal',
  password: 'Heslo1234',
}

export const TOKEN = {
  access_token: 'accessToken',
  token_type: 'bearer',
  expires_in: 7200,
  refresh_token: 'refreshToken',
  scope: 'market:335',
  created_at: 1555858950,
  owner_id: 1,
  owner_type: 'customer',
}

export const LOGIN_DATA = {
  id: USER.id,
  type: 'customers',
  links: {
    self: `https://the-amber-brand-12.commercelayer.io/api/customers/${
      USER.id
    }`,
  },
  attributes: {
    email: USER.email,
    status: 'prospect',
    created_at: '2019-04-21T15:02:29.631Z',
    updated_at: '2019-04-21T15:02:29.631Z',
    reference: null,
    metadata: { firstName: USER.firstName },
  },
}

export const ERROR_EMAIL_ALREADY_TAKEN = {
  title: 'has already been taken',
  detail: 'email - has already been taken',
  code: 'VALIDATION_ERROR',
  source: { pointer: '/data/attributes/email' },
  status: '422',
  meta: {
    error: 'taken',
    value: 'test@test.com',
  },
}
