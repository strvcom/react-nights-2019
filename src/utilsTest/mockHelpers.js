import fetchMock from 'fetch-mock'
import config from '../config'
import {
  PRODUCTS,
  PRODUCT,
  USER,
  LOGIN_DATA,
  TOKEN,
  ERROR_EMAIL_ALREADY_TAKEN,
} from './mockData'

export const mockUnauthorizedRequest = () =>
  fetchMock.restore().post(`${config.apiUrl}/oauth/token`, {
    status: 401,
    body: {},
  })

export const mockAuthorizedRequest = () =>
  fetchMock
    .restore()
    .post(`${config.apiUrl}/oauth/token`, {
      status: 200,
      body: TOKEN,
    })
    .get(`${config.apiUrl}/api/customers/${USER.id}`, {
      status: 200,
      body: {
        data: LOGIN_DATA,
      },
    })

export const mockFetchProducts = () =>
  mockUnauthorizedRequest().get(
    `${
      config.apiUrl
    }/api/skus?include=prices&page%5Bnumber%5D=1&page%5Bsize%5D=50`,
    {
      status: 200,
      body: PRODUCTS,
    }
  )

export const mockFetchProduct = () =>
  mockUnauthorizedRequest().get(`${config.apiUrl}/api/skus/1?include=prices`, {
    status: 200,
    body: PRODUCT,
  })

export const mockSignUp422 = () =>
  mockUnauthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 411,
    body: {
      errors: [ERROR_EMAIL_ALREADY_TAKEN],
    },
  })

export const mockSignUp201 = () =>
  mockAuthorizedRequest().post(`${config.apiUrl}/api/customers`, {
    status: 201,
    body: {
      data: LOGIN_DATA,
    },
  })
