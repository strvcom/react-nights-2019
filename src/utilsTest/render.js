import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from 'react-testing-library'

export const renderWithRouter = (component, uri) =>
  render(<MemoryRouter initialEntries={[uri]}>{component}</MemoryRouter>)
