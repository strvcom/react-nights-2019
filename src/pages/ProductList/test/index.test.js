import React from 'react'
import { waitForElement } from 'react-testing-library'
import 'jest-styled-components'

import { App } from '../../../App'
import * as routes from '../../../routes'
import { renderWithRouter } from '../../../utilsTest/render'
import { mockFetchProducts } from '../../../utilsTest/mockHelpers'
import { PRODUCTS } from '../../../utilsTest/mockData'

describe('[pages] ProductsList', () => {
  describe('when loading', () => {
    it('should render correctly', () => {
      const renderer = renderWithRouter(<App />, routes.PRODUCT_LIST)
      expect(renderer.container).toMatchSnapshot()
    })
  })

  describe('when products loaded', () => {
    mockFetchProducts()

    it('should render correctly', async () => {
      const renderer = renderWithRouter(<App />, routes.PRODUCT_LIST)
      await waitForElement(() => renderer.getByTestId('product-in-list'))
      expect(renderer.getAllByTestId('product-in-list').length).toEqual(
        PRODUCTS.data.length
      )
    })
  })
})
