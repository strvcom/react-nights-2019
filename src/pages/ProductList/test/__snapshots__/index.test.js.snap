// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[pages] ProductsList when loading should render correctly 1`] = `
.c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c4 {
  width: 100%;
  padding: 5rem 0;
  padding: 5rem 0;
}

.c5 {
  width: 5rem;
  height: 5rem;
  margin: 0 auto;
}

.c6 {
  -webkit-animation: cZxgpV 2s linear infinite;
  animation: cZxgpV 2s linear infinite;
  height: 100%;
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
  width: 100%;
}

.c7 {
  stroke: #ef0d33;
  stroke-dasharray: 1,200;
  stroke-dashoffset: 0;
  -webkit-animation: iwIXhf 1.5s ease-in-out infinite;
  animation: iwIXhf 1.5s ease-in-out infinite;
  stroke-linecap: round;
}

.c3 {
  font-family: circular,serif;
  font-weight: 400;
  color: #ef0d33;
  text-transform: uppercase;
  text-align: center;
}

<body>
  <div>
    <header
      class="c0"
    >
      <div
        class=""
      >
        <a
          class="c1"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class=""
      >
        <a
          class="c1"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="c1"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="c1"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="c2"
    >
      <h1
        class="c3"
      >
        E-Commerce app
      </h1>
      <div
        class="c4"
        data-testid="loader"
      >
        <div
          class="c5"
        >
          <svg
            class="c6"
            viewBox="25 25 50 50"
          >
            <circle
              class="c7"
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="4"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`[pages] ProductsList when products loaded should render correctly 1`] = `
<body>
  <div>
    <header
      class="sc-bwzfXH kuSkvH"
    >
      <div
        class="sc-htpNat KtFsv"
      >
        <a
          class="sc-bxivhb cHbCFb"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class="sc-htpNat KtFsv"
      >
        <a
          class="sc-bxivhb cHbCFb"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="sc-bxivhb cHbCFb"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="sc-bxivhb cHbCFb"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="sc-bdVaJa eGntFc"
    >
      <h1
        class="sc-htoDjs eTNtUt"
      >
        E-Commerce app
      </h1>
      .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style: none;
  padding: 0;
  text-align: center;
  width: 100%;
}

.c1 {
  margin: 5px;
}

<ul
        class="c0"
      >
        <li
          class="c1"
        >
          <a
            href="/products?page=1&size=25"
          >
            1
          </a>
        </li>
      </ul>
      <select>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
      .c7 {
  background: #ef0d33;
  cursor: pointer;
  padding: 1rem;
  margin-top: 0.5rem;
  border: none;
  border-radius: 5px;
  color: #fff;
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: initial;
  background-color: #fff;
  padding: 3rem 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  margin-bottom: 1rem;
  width: 35rem;
}

.c2 {
  height: 16rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  color: #ef0d33;
  font-size: 1.8rem;
  margin-top: 2rem;
}

.c3 {
  max-height: 15rem;
  max-width: 100%;
}

.c4 {
  height: 6.7rem;
  overflow: hidden;
}

.c5 {
  font-size: 1.6rem;
  font-weight: 100;
  text-transform: uppercase;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  margin-top: 4rem;
  padding: 0;
  list-style: none;
}

<ul
        class="c0"
      >
        <li
          class=""
        >
          <a
            class="c1"
            href="/products/1"
          >
            <div
              class="c2"
            >
              <img
                alt="First Test Product image"
                class="c3"
                src="https://img.commercelayer.io/skus/1.png"
              />
            </div>
            <div
              class="c4"
            >
              <h3
                class="c5"
              >
                First Test Product
              </h3>
            </div>
            <div
              class="c6"
            >
              $10
            </div>
            <button
              class="c7"
            >
              Add to Cart
            </button>
          </a>
        </li>
        <li
          class=""
        >
          <a
            class="c1"
            href="/products/2"
          >
            <div
              class="c2"
            >
              <img
                alt="Second Test Product image"
                class="c3"
                src="https://img.commercelayer.io/skus/1.png"
              />
            </div>
            <div
              class="c4"
            >
              <h3
                class="c5"
              >
                Second Test Product
              </h3>
            </div>
            <div
              class="c6"
            >
              $20
            </div>
            <button
              class="c7"
            >
              Add to Cart
            </button>
          </a>
        </li>
      </ul>
    </div>
  </div>
  .c2 {
  padding: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 0.1rem solid gainsboro;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 3rem;
}

.c1 {
  margin: 0 1rem;
}

.c3 {
  font-family: circular,serif;
  font-weight: 400;
  color: #ef0d33;
  text-transform: uppercase;
  text-align: center;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style: none;
  padding: 0;
  text-align: center;
  width: 100%;
}

.c5 {
  margin: 5px;
}

.c13 {
  background: #ef0d33;
  cursor: pointer;
  padding: 1rem;
  margin-top: 0.5rem;
  border: none;
  border-radius: 5px;
  color: #fff;
}

.c7 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: initial;
  background-color: #fff;
  padding: 3rem 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  margin-bottom: 1rem;
  width: 35rem;
}

.c8 {
  height: 16rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c12 {
  color: #ef0d33;
  font-size: 1.8rem;
  margin-top: 2rem;
}

.c9 {
  max-height: 15rem;
  max-width: 100%;
}

.c10 {
  height: 6.7rem;
  overflow: hidden;
}

.c11 {
  font-size: 1.6rem;
  font-weight: 100;
  text-transform: uppercase;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  margin-top: 4rem;
  padding: 0;
  list-style: none;
}

<div>
    <header
      class="c0"
    >
      <div
        class=""
      >
        <a
          class="c1"
          href="/products"
        >
          All Products
        </a>
      </div>
      <div
        class=""
      >
        <a
          class="c1"
          href="/cart"
        >
          My Cart
        </a>
        |
        <a
          class="c1"
          href="/login"
        >
          Log In
        </a>
         |
        <a
          class="c1"
          href="/signup"
        >
          Sign Up
        </a>
      </div>
    </header>
    <div
      class="c2"
    >
      <h1
        class="c3"
      >
        E-Commerce app
      </h1>
      <ul
        class="c4"
      >
        <li
          class="c5"
        >
          <a
            href="/products?page=1&size=25"
          >
            1
          </a>
        </li>
      </ul>
      <select>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
      <ul
        class="c6"
      >
        <li
          class=""
        >
          <a
            class="c7"
            href="/products/1"
          >
            <div
              class="c8"
            >
              <img
                alt="First Test Product image"
                class="c9"
                src="https://img.commercelayer.io/skus/1.png"
              />
            </div>
            <div
              class="c10"
            >
              <h3
                class="c11"
              >
                First Test Product
              </h3>
            </div>
            <div
              class="c12"
            >
              $10
            </div>
            <button
              class="c13"
            >
              Add to Cart
            </button>
          </a>
        </li>
        <li
          class=""
        >
          <a
            class="c7"
            href="/products/2"
          >
            <div
              class="c8"
            >
              <img
                alt="Second Test Product image"
                class="c9"
                src="https://img.commercelayer.io/skus/1.png"
              />
            </div>
            <div
              class="c10"
            >
              <h3
                class="c11"
              >
                Second Test Product
              </h3>
            </div>
            <div
              class="c12"
            >
              $20
            </div>
            <button
              class="c13"
            >
              Add to Cart
            </button>
          </a>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
